### DISCUSSION TITLE
Confusing about how to update pnpm itself

### DISCUSSION THREAD

**Discussion by wewindy** [pnpm/pnpm#4383] [Q&A] [answered]
ENV:

- Windows 10 21H2 x64
- install pnpm by using `iwr https://get.pnpm.io/install.ps1 -useb | iex`

---

One day I want to update pnpm itself to the latest version (for example, `6.31.0` → `6.32.1`), I run `pnpm add -g pnpm`, then everything seems OK, but after doing this I check the version again, it does not change?

``` sh
pnpm -v
6.31.0

pnpm add -g pnpm
# OK!

pnpm -v
6.31.0 # not change
```

![image](https://user-images.githubusercontent.com/26168955/155373895-3b3ad878-0092-4887-837b-0b8bf827cfbc.png)

**Comment by erwol** [comment, marked as answer]
2024 comment: the official docs now recommend using `pnpm self-update`: https://pnpm.io/installation#updating-pnpm

original 2022 comment: `pnpm add -g pnpm` should install latest available stable `pnpm`. You can cherry pick the version with `pnpm add -g pnpm@6.32.2` (latest stable at the moment https://github.com/pnpm/pnpm/releases).

**Reply by wewindy** [reply]
It seems ... oh~

![image](https://user-images.githubusercontent.com/26168955/156537712-9cd4c083-d7c8-4745-8fac-6391e58d1a41.png)

**Reply by cdaringe** [reply]
while it can work, i'd advise against it. what is described above is (paraphrased):

- install first pnpm via some means (brew, choco, deb, apk, iwr, whatever...)
- install second version of pnpm via pnpm

now you have two installs of pnpm, and can get into sticky/troublesome behavior.

node _ships_ with npm. 

i strongly recommend only installing `pnpm` _strictly via npm_, given that our community also uses tools like `nvm`/`fnm`/`asdf` _heavily_.

The following statement holds: "If you need pnpm, you also need node. If you have node installed, you have npm installed (in 99% cases), thus have an extremely reliable pnpm installation method, `npm i -g pnpm`". Using this reasoning as well, i only install _global_ packages using npm, as it's reliably present with 100% of my node versions. This argument is a bit fuzzier, because you lose some of pnpm's benefits there... so take that one with a grain of salt :)

**Reply by malgorzata-stefanowicz** [reply]
on point!!!

**Reply by pshergie** [reply]
doesn't work
<img width="477" alt="image" src="https://github.com/pnpm/pnpm/assets/14779553/b3f92f19-8ec0-41a3-b2c5-1c72380fcbf2">

**Reply by shadabairbase** [reply]
Thanks @cdaringe your comment helped me.
I am on m3(arm64) and my Node gets managed by nvm.
I ran the below command to check about the multiple pnpm installation and find out that indeed I had two installtions.
```
which -a pnpm
type -a pnpm
```
Output
```
/Users/bond007/.nvm/versions/node/v20.10.0/bin/pnpm
/Users/bond007/Library/pnpm/pnpm
/Users/bond007/.nvm/versions/node/v20.10.0/bin/pnpm
pnpm is /Users/bond007/.nvm/versions/node/v20.10.0/bin/pnpm
pnpm is /Users/bond007/Library/pnpm/pnpm
pnpm is /Users/bond007/.nvm/versions/node/v20.10.0/bin/pnpm
```

So I uninstalled the pnpm version managed by nvm
```
npm uninstall -g pnpm
```
And then I was able to get the right pnpm version.

**Reply by lukeMWI** [reply]
@Erwol Worked for me, thanks! I was going from 8.15.4 -> 9.1.4 on a Mac with the Intel chip

**Reply by Arcitec** [reply]
I agree with @cdaringe! I was using `sudo corepack enable` to install pnpm long ago, but this is incapable of updating pnpm. I tried using `sudo corepack prepare pnpm@latest --activate` but it never gave me anything newer than PNPM from 2 years ago.

So I went `sudo corepack disable` and switched to installing PNPM via NPM instead:

`sudo npm i -g pnpm`

Now the chain is: System's Node/NPM packages manage my PNPM installation, and PNPM manages all my projects and all my other globally installed tools. No more problems updating PNPM itself! It's simple: `sudo npm update -g pnpm` :)

**Reply by CogSciOliver** [reply]
This worked thanks a million!
My steps for reference: 

* First 
`pnpm self-update`
* Updated but not active from -v check  

`pnpm self-update`
* Latest was added but, not active

`npm uninstall -g pnpm`
* Block removed 

`sudo npm i -g pnpm`
* NPM switch made 

`pnpm -v`
* Confirmed latest active

**Comment by yingpengsha** [comment]
`npm install pnpm -g`

**Reply by pimboden** [reply]
Important.... npm  NOT pnpm install...  If you first installed pnpm with npm. you'll hae to use npm

**Reply by justingolden21** [reply]
This worked and `pnpm add -g pnpm` didn't.

**Reply by SSylvain1989** [reply]
work fine for me on macOS ( intel ) 
i was trying 
```
pnpm add -g pnpm
Nothing to stop. No server is running for the store at ...
```
i was on :
`
pnpm -v
8.6.0
`
now with `npm install pnpm -g`

```
npm install pnpm -g

changed 1 package, and audited 2 packages in 3s

1 package is looking for funding
  run `npm fund` for details

found 0 vulnerabilities

pnpm -v
8.15.2
```

**Reply by kuldip-simform** [reply]
Only this command worked for me.

**Reply by arantebw** [reply]
In Fedora/Linux, this one worked for me:

```bash
sudo npm install -g pnpm
```

**Reply by douglasg14b** [reply]
This seems to result in:

```
npm error code EEXIST
npm error path /home/douglas/.nvm/versions/node/v20.12.2/bin/pnpm
npm error EEXIST: file already exists
npm error File exists: /home/douglas/.nvm/versions/node/v20.12.2/bin/pnpm
```

**Comment by yanggggjie** [comment]
try "pnpm add -g @pnpm/exe", it works for me!
![image](https://user-images.githubusercontent.com/60206592/202073002-fcc4351b-2a66-49ab-a894-c16400ac116a.png)

![image](https://user-images.githubusercontent.com/60206592/202072927-ee3b33c6-82ab-4de6-a90d-b80e19466722.png)

**Reply by heirro** [reply]
Thanks, it's worked on my Mac M1

**Reply by miraclemenikelechi** [reply]
just tried, did not work. `npm i pnpm -g` worked tho

**Comment by dmZhan** [comment]
> ENV:
> 
> * Windows 10 21H2 x64
> * install pnpm by using `iwr https://get.pnpm.io/install.ps1 -useb | iex`
> 
> One day I want to update pnpm itself to the latest version (for example, `6.31.0` → `6.32.1`), I run `pnpm add -g pnpm`, then everything seems OK, but after doing this I check the version again, it does not change?
> 
> ```shell
> pnpm -v
> 6.31.0
> 
> pnpm add -g pnpm
> # OK!
> 
> pnpm -v
> 6.31.0 # not change
> ```
> 
> ![image](https://user-images.githubusercontent.com/26168955/155373895-3b3ad878-0092-4887-837b-0b8bf827cfbc.png)

 I used the same method to install pnpm，and i used 'iwr https://get.pnpm.io/install.ps1 -useb | iex' in my PowerShell to update pnpm, I did it.
![1](https://user-images.githubusercontent.com/37793739/218469157-0ed461ab-59a8-411f-abc3-8cfb6f98e209.PNG)
![2](https://user-images.githubusercontent.com/37793739/218469208-6ff155e8-0909-4aa5-b42e-317082d3ed55.png)
![3](https://user-images.githubusercontent.com/37793739/218469292-52d6ff9f-90aa-4cef-8065-80dcd4dac8fd.PNG)
You must use 'iwr https://get.pnpm.io/install.ps1 -useb | iex' in **PowerShell** to update pnpm.

**Comment by benface** [comment]
If you installed `pnpm` via `corepack`, you can do: `corepack prepare pnpm@latest --activate`

**Reply by iobuhov** [reply]
Always updating/installing through `corepack` and never have troubles, the only thing to keep in mind -- if you use `nvm` you have to install pnpm after updating your node version

**Reply by dtinth** [reply]
The command is now:

```sh
corepack install --global pnpm
```

**Reply by stormwatch** [reply]
> The command is now:
> 
> ```shell
> corepack install --global pnpm
> ```

Why the downvotes? Both commands will upgrade pnpm, only to different versions:
`corepack prepare pnpm@latest --activate` upgraded from 9.15.2 (the version reported after `corepack enable pnpm` in node v23.6.0), to v9.15.4

`corepack install --global pnpm` upgraded pnpm from v9.15.4 to v10.0.0

**Reply by Arcitec** [reply]
@stormwatch I guess that explains why `prepare` always gave me a several years outdated version of pnpm seen here:

https://github.com/orgs/pnpm/discussions/4383#discussioncomment-11133513

**Comment by joan17cast** [comment]
The solution for me was remove pnpm manually and reinstall. To remove manually i use the command `where pnpm` on the windows terminal to reveal all the locations where is installed the pnpm package and then remove manually. Once you have this done simply reinstall pnpm using this command `iwr https://get.pnpm.io/install.ps1 -useb | iex` and success 🎉.

**Comment by JuanAntonioV** [comment]
I have a same problem couple weeks ago, i fixed it by running `npm install -g pnpm` or `npm install -g @pnpm/exe` instead of  `pnpm i -g pnpm`. The main things here, you cannot install or updating pnpm by pnpm it self. but you can use npm instead. I hope it help.

**Comment by supermacro** [comment]
Would be great to have a built-in command to have pnpm update. Much like bun has `bun upgrade`

**Comment by tgrushka** [comment]
Wow, why don't they just add:

Try `pnpm i -g pnpm` or `npm i -g pnpm`

to these instructions ("Run a script from (blah blah blah)" + having to visit that page and read through a whole bunch of scenarios is just not user friendly):
```
   ╭─────────────────────────────────────────────────────────────────╮
   │                                                                 │
   │                Update available! 8.8.0 → 8.9.2.                 │
   │   Changelog: https://github.com/pnpm/pnpm/releases/tag/v8.9.2   │
   │         Run a script from: https://pnpm.io/installation         │
   │                                                                 │
   │     Follow @pnpmjs for updates: https://twitter.com/pnpmjs      │
   │                                                                 │
   ╰─────────────────────────────────────────────────────────────────╯

```

And yes, `pnpm i -g pnpm` does work on macOS if `pnpm` was originally installed via `npm`.

**Reply by Devanshukoli** [reply]
Yeah. Absolutely.

**Comment by alipalvane** [comment]
i have same problem
just go directory or path in WARN and remove pnpm.exe and again run `npm install -g pnpm` or `pnpm install -g pnpm`
its work for me

**Comment by Rednas83** [comment]
**Easiest is to install pnpm with corepack**
`corepack enable pnpm` [with adminstrator rights from any path]
`corepack use pnpm@latest` [to manage the pnpm package in package.json via the experimental "packageManager" from the project path]

**Example of package.json**
`"packageManager": "pnpm@8.15.1+sha256.245fe901f8e7fa8782d7f17d32b6a83995e2ae03984cb5b62b8949bfdc27c7b5"`

I guess when there is an update you will need to run `corepack use pnpm@latest` again but perhaps there is also a way to update from **package.json** itself🤔

**Reply by vicary** [reply]
This is exactly what I needed, thanks a lot.

**Comment by dei8bit** [comment]
One more answer:  😋

`iwr https://get.pnpm.io/install.ps1 -useb | iex`

_done_

**Comment by zkochan** [comment]
I am adding a new command to make it easier: https://github.com/pnpm/pnpm/pull/8424

**Reply by patarapolw** [reply]
Terminal simply says

> Update available! 9.6.0 → 9.10.0.
> Changelog: https://github.com/pnpm/pnpm/releases/tag/v9.10.0
> Run a script from: https://pnpm.io/installation  

And the website on the last line says

> ## Updating pnpm
>
> To update pnpm, run the [self-update](https://pnpm.io/cli/self-update) command:
>
> ```
> pnpm self-update
> ```

While actually running `self-update` doesn't work (not yet).

```
> pnpm self-update
'self-update' is not recognized as an internal or external command,
operable program or batch file.
 ERR_PNPM_RECURSIVE_EXEC_FIRST_FAIL  Command "self-update" not found
```

So, I just have to run the above command (but not in docs).

```
iwr https://get.pnpm.io/install.ps1 -useb | iex
```

Only after update that the `self-update` command works.

I can't remember whether I installed `pnpm` via `iex` or via `corepack`, but it's more probably the former.

**Reply by Ali559** [reply]
This doesn't work on macos with m1 chip: 
```
> pnpm self-update
 ERR_PNPM_CANT_SELF_UPDATE_IN_COREPACK  You should update pnpm with corepack
```

**Reply by erwol** [reply]
@Ali559 this was fixed here https://github.com/pnpm/pnpm/pull/8544; but to get the fix you'll need to get the latest pnpm... you will probably have to fully reinstall your pnpm.

**Reply by lifeforced** [reply]
On pnpm 9.15.1 installed using corepack running `pnpm self-update` gives an error saying to use corepack to update pnpm. I wonder if the `self-update` command can be modified to automatically do this?

**Comment by hoiles** [comment]
I have found out that you need to update it, using the same package that you used for the installation
For example, in my case, I tried 
`pnpm self-update` but since my current version was 8.13, It could not be updated[only in version [9](https://pnpm.io/cli/self-update)]

Since I could not remember how I installed pnpm then I check if I did it with NPM 
`npm ls —dept=0` 
to list the dependencies installed, but unfortunately, the pnpm was not there.

Then I checked homebrew

`brew list`

And there it was my pnpm in the outdated formulae packages 😭 

I just run 

`brew install pnpm`
`brew update`

And voila, I am now in 9.11 😄 . Can't wait now to have a newer version to try the `self-update`

**Reply by brunobely** [reply]
For others in this situation (on macOS), I checked that `pnpm` had not been installed with either `npm` or `brew`, so figured I probably used the install script from https://pnpm.io/installation

```
curl -fsSL https://get.pnpm.io/install.sh | sh -
```

Which worked to update it.

**Comment by TonyGravagno** [comment]
Running pnpm v9.10.0 earlier today, I used npm to update pnpm to 9.12.1. But the Windows 10 path wasn't updated even after a reboot. So the system `pnpm -v` continues to show 9.10.0. I'm not sure how this worked but I used `pnpm self-update` later today and it installed 9.12.2, but still didn't update the path. I can go to the 9.12.2 folder and execute and get the correct version:

> C:\Users\tonyg\AppData\Local\pnpm\.tools\pnpm\9.12.2\node_modules\.pnpm\pnpm@9.12.2\node_modules\pnpm\bin>node .\pnpm.cjs
> Version 9.12.2 (compiled to binary; bundled Node.js v22.6.0)

Can someone provide a tip for how this should be globally enabled/updated?

Side note: "self-update" wasn't added to the usage info yet.

Thanks for this great update!

**Comment by alessandroperetti** [comment]
I had an old `pnpm` version. 
In order to update it I followed these steps:

1. `npm uninstall -g pnpm`
3. `npm i -g pnpm`
4. `pnpm --version`

**Comment by robsterlini** [comment]
Was hoping to be able to use `pnpm self-update` in tandem with `engines.pnpm`, but I get this error:

```
➜  redacted-XXX git:(robsterlini/update-pnpm) ✗ pnpm self-update
 ERR_PNPM_UNSUPPORTED_ENGINE  Unsupported environment (bad pnpm and/or Node.js version)

Your pnpm version is incompatible with "redacted-XXX".

Expected version: 9.15.3
Got: 9.12.3

This is happening because the package's manifest has an engines.pnpm field specified.
To fix this issue, install the required pnpm version globally.

To install the latest version of pnpm, run "pnpm i -g pnpm".
To check your pnpm version, run "pnpm -v".
```

9.12.3 is high enough that `self-update` exists, but the engines check means it fails. Is there a way around this?

**Comment by soundharsaminathan** [comment]
This works for windows `Invoke-WebRequest https://get.pnpm.io/install.ps1 -UseBasicParsing | Invoke-Expression`
https://pnpm.io/installation#on-windows

**Comment by Ultracrepidarianism** [comment]
It seems like the answer can also depend on the package manager you used to install pnpm. I used chocolatey and nothing worked except 
`choco upgrade pnpm`

**Comment by laphilosophia** [comment]
Here is the solution I’ve applied myself for Windows users:

The error occurs because Windows prevents deleting a file that is currently in use, such as `pnpm.exe` when the `pnpm` process is running.  To resolve this, you must rename the existing `pnpm.exe` file before running the update command. 

Recommended Workaround Steps:

- Locate `pnpm.exe` in `*\*\*\AppData\Local\pnpm\`.
- Rename `pnpm.exe` to `pnpm_old.exe` (or similar).
- Run the update command using the renamed binary: `pnpm_old.exe -g i pnpm`.
- Once the new version installs successfully, delete the old `pnpm_old.exe` file.

**Comment by kingdanie** [comment]
I ran into this same issue on my Mac (M series chip). I tried running `pnpm self-update`, but `pnpm -v` still showed the old version.

**The reason:**
If you installed pnpm using Homebrew, the self-update command won't work correctly. This is because Homebrew "owns" the pnpm file in a specific folder, and even if pnpm downloads an update to a different folder, your Mac will keep using the Homebrew version.

**The solution:**
If you used Homebrew to install pnpm, you must use Homebrew to update it. Run this command instead:

```
brew upgrade pnpm
```

**Comment by dei8bit** [comment]
omg this continue ? xD it's time to switch to `bun` my friends
